wire-ios-sync-engine/Source/Synchronization/RecurringActionService.swift

```diff
@@ -0,0 +1,95 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+struct RecurringAction {
+
+    typealias Action = () -> Void
+
+    let id: String
+    let interval: TimeInterval
+    let perform: Action
+
+}
+
+protocol RecurringActionServiceInterface {
+
+    func performActionsIfNeeded()
+    func registerAction(_ action: RecurringAction)
+
+}
+
+class RecurringActionService: RecurringActionServiceInterface {
+
+    private var actions = [RecurringAction]()
+
+    public func performActionsIfNeeded() {
+        let currentDate = Date()
+
+        actions.forEach { action in
+
+            guard let lastActionDate = lastActionDate(for: action.id) else {
+                persistLastActionDate(for: action.id, newDate: currentDate)
+                return
+            }
+
+            if (lastActionDate + action.interval) <= currentDate {
+                action.perform()
+                persistLastActionDate(for: action.id, newDate: currentDate)
+            }
+
+        }
+    }
+
+    public func registerAction(_ action: RecurringAction) {
+        actions.append(action)
+    }
+
+    // MARK: - Helpers
+
+    func lastActionDate(for actionID: String) -> Date? {
+        return userDefaults(for: actionID)?.lastRecurringActionDate
+    }
+
+    func persistLastActionDate(for actionID: String, newDate: Date) {
+        userDefaults(for: actionID)?.lastRecurringActionDate = newDate
+    }
```

Comment:
Do you need to pass in the new date if it is always `Date()`?

Reply:
it's useful for tests

Reply:
and it may take some time for the action to run, and I think it's good to persist the same date that we use to execute the actions.

Reply:
I removed the date from the method parameters.